akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -55,10 +55,23 @@ private[http] object HttpServerBluePrint {
       }.withDeploy(Deploy.local)
     }, errorMsg = "Http.serverLayer is currently not reusable. You need to create a new instance for each materialization.")
 
+    def establishAbsoluteUri(requestOutput: RequestOutput): RequestOutput = requestOutput match {
+      case start: RequestStart ⇒
+        try {
+          val effectiveUri = HttpRequest.effectiveUri(start.uri, start.headers, securedConnection = false, defaultHostHeader)
+          start.copy(uri = effectiveUri)
+        } catch {
+          case e: IllegalUriException ⇒
+            MessageStartError(StatusCodes.BadRequest, ErrorInfo("Request is missing required `Host` header", e.getMessage))
```

Comment:
It depends on the perspective: if you haven't defined a default host, it could mean that you just don't want to support HTTP/1.0 in which case 400 seems appropriate. Or you see it as a server side misconfiguration leading to a 500 which seems paradox because why would we allow such a misconfiguration in the first place?